### PR TITLE
PP-13051 Introduce missing test combinations

### PR DESCRIPTION
**Pull request by marcotranchino**
With previous work, we have introduced corporate exemptions and tests to ensure the correct behaviour depending on:

- exemption engine being ON or OFF
- corporate exemptions being ON or OFF
- cards being CORPORATE or NON CORPORATE

With this change, we are adding three new tests for the three missing combinations as follows:

- exemption engine ON, corporate exemptions ON, NON corporate card --> OPTIMISED REQUESTED
- exemption engine OFF, corporate exemptions ON, corporate card --> CORPORATE REQUESTED
- exemption engine OFF, corporate exemptions ON, NON corporate card --> NOT REQUESTED

Further information in the JIRA ticket.

https://payments-platform.atlassian.net/browse/PP-13051

# TEST COMBINATIONS

test combinations

Exemption engine | Corporate Exemptions | Corporate card | Result | Status | Name
-- | -- | -- | -- | -- | --
TRUE | TRUE | TRUE | CORPORATE REQUESTED | existing | should_send_corporate_exemption_with_worldpay_authorisation_if_corporate_exemption_enabled_and_exemption_engine_enabled_and_corporate_card_used
TRUE | TRUE | FALSE | OPTIMISED REQUESTED | new | should_send_optimised_exemption_with_worldpay_authorisation_if_corporate_exemption_enabled_and_exemption_engine_enabled_and_non_corporate_card_used
TRUE | FALSE | TRUE | OPTIMISED REQUESTED | existing | should_send_optimised_exemption_with_worldpay_authorisation_if_corporate_exemption_is_not_enabled_and_exemption_engine_enabled_and_corporate_card_used
TRUE | FALSE | FALSE | OPTIMISED REQUESTED | existing | should_include_exemption_if_account_has_exemption_engine_set_to_true
FALSE | TRUE | TRUE | CORPORATE REQUESTED | new | should_send_corporate_exemption_with_worldpay_authorisation_if_corporate_exemption_enabled_and_exemption_engine_disabled_and_corporate_card_used
FALSE | TRUE | FALSE | NOT REQUESTED | new | should_not_request_exemption_if_exemption_engine_disabled_and_corporate_exemptions_enabled_and_non_corporate_card_used
FALSE | FALSE | TRUE | NOT REQUESTED | existing | should_not_send_corporate_exemption_with_worldpay_authorisation_if_corporate_exemptions_false_and_corporate_card_is_used
FALSE | FALSE | FALSE | NOT REQUESTED | existing | should_not_include_exemption_if_account_has_exemption_engine_set_to_false

### CODE REVIEW

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java [hunk 1]
```diff
@@ -384,6 +385,40 @@ void should_not_include_exemption_if_account_has_exemption_engine_set_to_true_bu
...
+
+        ChargeEntity chargeEntity = chargeEntityFixture.build();
+        var cardAuthRequest = new CardAuthorisationGatewayRequest(chargeEntity, anAuthCardDetails().withCorporateCard(false).build());
+
+        when(worldpayAuthoriseHandler.authorise(cardAuthRequest, DO_NOT_SEND_EXEMPTION_REQUEST))
+                .thenReturn(getGatewayResponse(WORLDPAY_AUTHORISATION_SUCCESS_RESPONSE));
+
+        worldpayPaymentProvider.authorise(cardAuthRequest, chargeEntity);
+
+        assertThat(chargeEntity.getExemption3dsRequested(), is(nullValue()));
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java:408]: For completeness, can we assert the value of `chargeEntity.getExemption3ds()` too?

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java [hunk 2]
```diff
@@ -510,6 +545,76 @@ void should_send_corporate_exemption_with_worldpay_authorisation_if_corporate_ex
...
+        var cardAuthRequest = new CardAuthorisationGatewayRequest(chargeEntity, anAuthCardDetails()
+                .withCorporateCard(true).build());
+
+        when(chargeDao.merge(chargeEntity)).thenReturn(chargeEntity);
+        when(worldpayAuthoriseHandler.authorise(cardAuthRequest, SEND_CORPORATE_EXEMPTION_REQUEST))
+                .thenReturn(getGatewayResponse(WORLDPAY_EXEMPTION_REQUEST_HONOURED_RESPONSE));
+
+        worldpayPaymentProvider.authorise(cardAuthRequest, chargeEntity);
+
+        assertThat(chargeEntity.getExemption3dsRequested(), is(CORPORATE));
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: I think the line below covers this.

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java [hunk 3]
```diff
@@ -510,6 +545,76 @@ void should_send_corporate_exemption_with_worldpay_authorisation_if_corporate_ex
...
+        var cardAuthRequest = new CardAuthorisationGatewayRequest(chargeEntity, anAuthCardDetails()
+                .withCorporateCard(false).build());
+
+        when(chargeDao.merge(chargeEntity)).thenReturn(chargeEntity);
+        when(worldpayAuthoriseHandler.authorise(cardAuthRequest, SEND_EXEMPTION_ENGINE_REQUEST))
+                .thenReturn(getGatewayResponse(WORLDPAY_EXEMPTION_REQUEST_HONOURED_RESPONSE));
+
+        worldpayPaymentProvider.authorise(cardAuthRequest, chargeEntity);
+
+        assertThat(chargeEntity.getExemption3dsRequested(), is(OPTIMISED));
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: Again, the line below covers this.

**Review by alexbishop1 — CHANGES_REQUESTED** [review]
I like a PR with a table!

This looks good, just a couple of small suggestions.

It’s a shame the tests have so much boilerplate but what can you do.

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java (hunk 2 above)
Comment (marcotranchino) [reply src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: Amended

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java (hunk 3 above)
Comment (marcotranchino) [reply src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: Amended

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java (hunk 1 above)
Comment (marcotranchino) [reply src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java:408]: Done.

**Comment by marcotranchino** [comment]
Thanks, Alex! 
Feedback applied :)

**Review by alexbishop1 — APPROVED** [review]

**Pull request merged**